pkg/apis/sources/v1beta1/apiserver_types.go

```diff
@@ -0,0 +1,149 @@
+
+// ApiServerSource is the Schema for the apiserversources API
+type ApiServerSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ApiServerSourceSpec   `json:"spec,omitempty"`
+	Status ApiServerSourceStatus `json:"status,omitempty"`
+}
+
+// Check the interfaces that ApiServerSource should be implementing.
+var (
+	_ runtime.Object     = (*ApiServerSource)(nil)
+	_ kmeta.OwnerRefable = (*ApiServerSource)(nil)
+	_ apis.Validatable   = (*ApiServerSource)(nil)
+	_ apis.Defaultable   = (*ApiServerSource)(nil)
+	_ apis.HasSpec       = (*ApiServerSource)(nil)
+)
+
+const (
+	// ApiServerSourceAddEventType is the ApiServerSource CloudEvent type for adds.
+	ApiServerSourceAddEventType = "dev.knative.apiserver.resource.add"
+	// ApiServerSourceUpdateEventType is the ApiServerSource CloudEvent type for updates.
+	ApiServerSourceUpdateEventType = "dev.knative.apiserver.resource.update"
+	// ApiServerSourceDeleteEventType is the ApiServerSource CloudEvent type for deletions.
+	ApiServerSourceDeleteEventType = "dev.knative.apiserver.resource.delete"
+
+	// ApiServerSourceAddRefEventType is the ApiServerSource CloudEvent type for ref adds.
+	ApiServerSourceAddRefEventType = "dev.knative.apiserver.ref.add"
+	// ApiServerSourceUpdateRefEventType is the ApiServerSource CloudEvent type for ref updates.
+	ApiServerSourceUpdateRefEventType = "dev.knative.apiserver.ref.update"
+	// ApiServerSourceDeleteRefEventType is the ApiServerSource CloudEvent type for ref deletions.
+	ApiServerSourceDeleteRefEventType = "dev.knative.apiserver.ref.delete"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ApiServerSourceList contains a list of ApiServerSource
+type ApiServerSourceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ApiServerSource `json:"items"`
+}
+
+// ApiServerSourceSpec defines the desired state of ApiServerSource
+type ApiServerSourceSpec struct {
+	// inherits duck/v1 SourceSpec, which currently provides:
+	// * Sink - a reference to an object that will resolve to a domain name or
+	//   a URI directly to use as the sink.
+	// * CloudEventOverrides - defines overrides to control the output format
+	//   and modifications of the event sent to the sink.
+	duckv1.SourceSpec `json:",inline"`
+
+	// WatchResources is the list of resources to watch.
+	WatchResources []ApiServerResource `json:"watch"`
+
+	// ServiceAccountName is the name of the ServiceAccount to use to run the
+	// receive adapter for this source.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// Mode is the mode the receive adapter controller runs under: Reference or Resource.
+	// `Reference` sends only the reference to the resource that had a lifecycle event.
+	// `Resource` send the full resource lifecycle event.
+	// Defaults to `Reference`
+	// +optional
+	Mode string `json:"mode,omitempty"`
```

Comment:
Just a thought, would it be useful/feasible to control the Mode based on the resources being watched? If you want to get refs for some types of events but full resource for others, would it make sense to be able to run that at the resource level?

Reply:
And reading little further down, I'm actually a bit confused. Seems like this flag controls if you're sending a reference or the actual resource, unless you're setting the ResourceOwner in which case you're actually in some cases maybe getting a reference to the actual owner.
And I think the comment is wrong because it's not always the resource that had the lifecycle event, but sometimes it could be the owner, or maybe I'm just confused?

Reply:
Maybe instead of having the Mode be: Reference, Resource it should be: Reference, Resource, Owner
And if you then move it down below where it's at the definition of what to watch, it would make more sense. You'd then define the:
Filter (kind,apiversion,labelselector, ownerref)
How: Reference, Resource, Owner
there?

Reply:
The mode is data-ref vs data.

so if foo/barr changes as mode Reference: you get a data that just says "changed for foo/bar detected"

and if it is Resource: you get the full update of foo/bar as a data payload.

Reply:
I think of this more as `Verbose` vs `Pointer` or something like that.

Reply:
What do you think of using dataref extension? https://github.com/cloudevents/spec/blob/master/extensions/dataref.md

So we can have two modes: ["data" and "dataref"] or ["Full" and "ClaimCheck"] or something that makes sense.